### PR TITLE
Feature/util scripts

### PR DESCRIPTION
**Pull request by CrepeGoat**
This is meant to be a potential method for implementing #44 (which would address #36). I'm leaving this as a draft for discussion. This is for the most part untested.

### CODE REVIEW

File: ebmlite/cli.py
Comment (StokesMIDE) [inline ebmlite/cli.py:21]: Using a context here seems overly complex; I'm not sure it adds anything. There's only one schema per run of the script, one input file, and at most one output file. The context manager is cool but I think it makes things harder to read.

**Review by StokesMIDE — CHANGES_REQUESTED** [review]
The name `cli.py` could be more descriptive. Something like `cli_scripts` or `cli_utilities` or `utility_scripts` might be better.

File: ebmlite/cli.py
Comment (StokesMIDE) [inline ebmlite/cli.py:127]: "View" doesn't generate EBML.

File: ebmlite/cli.py
Comment (StokesMIDE) [inline ebmlite/cli.py:39]: Why not make 'mode' a parameter, and just supply either `"w"` or `"wb"` as the argument?

Comment (CrepeGoat) [reply ebmlite/cli.py:39]: because it removes the possibility of misusing the function with `mode='r'`, or god forbid `mode=5`.

why should I make a mode parameter? brevity?

File: ebmlite/cli.py
Comment (CrepeGoat) [reply ebmlite/cli.py:21]: failing to clean up file handles is a bug, and it's not immediately identifiable (esp. with our tests). I don't trust myself to remember to explicitly put in a proper exception-proof cleanup step in every function where I use this because I am, on the whole, pretty error-prone. Context managers are the safest guarantee I have for proper file handle management. If you want to change it, I will leave it to you to do so safely and correctly.

File: setup.py
Comment (StokesMIDE) [inline setup.py:42]: Should these reference `ebmlite.cmdline_script.py` instead of `ebmlite.cli`? 

Maybe just change `cmdline_scripts.py` back to `cli.py`. I misunderstood how it worked before; I didn't notice the `entry_points` in `setup.py`, and I thought the filename was exposed to the user.

**Comment by CFlaniganMide** [comment]
Has this been made redundant by [util.py fixes #70](https://github.com/MideTechnology/ebmlite/pull/70)?

**Comment by StokesMIDE** [comment]
> Has this been made redundant by [util.py fixes #70](https://github.com/MideTechnology/ebmlite/pull/70)?

They're different; this one addresses the way the utilities are structured, #70 adds a feature (generating XML without the extra attributes). The text/binary file fix in #70 is probably redundant, though.

Aside: The #70 change adds a `--min` command line argument to exclude the extra attributes that are usually unnecessary. If the way the utility is used is going to be changed anyway, I think it would be better to have the extra XML attributes excluded by default, with arguments to explicitly add each (ID, offset, size, and maybe `--max` as a shortcut for adding them all).

**Comment by CrepeGoat** [comment]
As noted by @StokesMIDE, a design requirement for this change is that the file is still runnable as a module (e.g. via `python /.../util.py ...`). In terms of implementation, this should be doable as long as the file still contains a functional `if __name__ == "__main__"` clause that runs all the same functions.

**Comment by CrepeGoat** [comment]
> a design requirement for this change is that the file is still runnable as a module... In terms of implementation, this should be doable...

...on second look, it's a little involved since I changed each utility function to parse its own args; I'd have to revert that change to get back the `if __name__ == "__main__"` functionality.

@StokesMIDE why did you want that exactly? Specifically, when would someone be running `ebmlite` code w/o `pip`-installing the repo?

**Comment by StokesMIDE** [comment]
> ...on second look, it's a little involved since I changed each utility function to parse its own args; I'd have to revert that change to get back the `if __name__ == "__main__"` functionality.

Each utility function would be its own eponymous Python file: `xml2ebml.py`, `ebml2xml.py`, etc. It doesn't significantly affect the implementation, it just moves some things around.

> @StokesMIDE why did you want that exactly? Specifically, when would someone be running `ebmlite` code w/o `pip`-installing the repo?

During development, or if they want to test or use (but not commit to installing) a specific branch, or don't want to modify their existing install. Pip can sometimes be a headache when trying to install versions over other versions, particularly if one was installed from PyPI, and/or if version numbers don't change.

**Comment by CrepeGoat** [comment]
> Each utility function would be its own eponymous Python file: `xml2ebml.py`, `ebml2xml.py`, etc. 

oh! yeah if you don't mind them being different files (as opposed to all one file, as it's currently written in development), then yeah that's easy I can do that.

**Comment by StokesMIDE** [comment]
What about `tools` as the directory name instead of `console_scripts`? I'm not sure it's a standard per se, but I've seen some other packages (like PySerial) organize like that.

**Comment by CrepeGoat** [comment]
uhhhm yeah they're not *just* console scripts, are they? sure `tools` works.

**Comment by StokesMIDE** [comment]
What was the intention behind the modular modularization in commit 9e7374802d3b700f6b17a05df1406e767cf4b4f9 ? I'm not sure things needed to be broken up so much. The granularity and uniform function names make it a bit hard to follow.

**Comment by CrepeGoat** [comment]
The intention was to use better coding practices in general, but specifically in this case it allows us to call the actual script from within Python (as we might want to do in a test) w/ something like e.g. `import ebmlite.tools.ebml2xml; ebmlite2xml.run(...)`, and w/o having to hack it by:
1. stubbing the `argparse` methods/object, or
2. forcing us to actually call the functions through the command line.

If you would prefer I can just revert it; it's not super important to me.

**Comment by CrepeGoat** [comment]
Note: ideally I'd like to merge this in **after #80** to ensure the original `util.py` hasn't changed throughout this PR.

**Comment by StokesMIDE** [comment]
> The intention was to use better coding practices in general, but specifically in this case it allows us to call the actual script from within Python (as we might want to do in a test) w/ something like e.g. `import ebmlite.tools.ebml2xml; ebmlite2xml.run(...)`, and w/o having to hack it by:
> 
>     1. stubbing the `argparse` methods/object, or
>     2. forcing us to actually call the functions through the command line (though tbh I'd kind of prefer doing so for testing in this case).
> 
> 
> If you would prefer I can just revert it; it's not super important to me.

I see your point about having the `run()` separated from the argument parsing, but splitting things into `formatArgs()` and the various `parseArgs()` and `main()` functions seem a bit too fine-grained. It would probably be better just to put that functionality into each utility's `main()`;  everything but `parseArgs()` is only a couple of lines long. 

For now, though, maybe just revert it rather than refactor it, just to keep the PR rolling.

**Review by StokesMIDE — APPROVED** [review]
Looks good. We can revisit re-modularizing things later.

**Comment by CrepeGoat** [comment]
yeah I'd still like to have the tests in from #80 first. ...which is unfortunate because I think this would best go on DAQ-maintenance and I'm already over on hours for that :\

**Review by StokesMIDE — APPROVED** [review]

**Pull request merged**
